polar-core/src/partial/partial.rs

```diff
@@ -1007,6 +1023,17 @@ mod test {
                i(y: C) if y.c > 0;
                i(y: B{bar: 2}) if y.b > 0;
 
+               # traversing `in` is tough!
+               r(a: A) if b in a.b and t(b);
+               s(a: A) if b in a.b and b matches B and u(b.c);
+               # This is still incorrect
+               # this should be equivalent to `s`
+               s_bad(a: A) if b in a.b and u(b.c);
```

Comment:
This is only a partial fix. If we don't know the type of `b`, then we can't do the isa with path check to find if `b.c matches C`. The problem is that the partial will look like: `_this matches A{} and b in _this.b` and we are checking `b.c matches C{}`. But we won't be able to "traverse" that `in` to check if `a.b.c matches C`. 

On the other hand, it'll work fine for `b in _this.b` checking if `b matches B{}`, because the new logic here will substitute `_this.b` for `b`. and so check whether `_this.b matches B`. Which currently works in the sqlalchemy logic.